lib/trento/application/process_managers/deregistration_process_manager.ex

```diff
@@ -19,84 +32,295 @@ defmodule Trento.DeregistrationProcessManager do
+      HostRegistered event.
+    - "Rolled-Up" events:
+        - HostRolledUp as the HostRegistered event might have been rolled up.
+        - ClusterRolledUp as the HostAddedToCluster event might have been rolled up.
+        - SapSystemRolledUp as the DatabaseInstanceRegistered/ApplicationInstanceRegistered events might have been
+          rolled up.
 
-    The process manager starts with a Deregistration request and stops when the host is fully deregistered.
+    HostDeregistered stops a Process Manager for the Host identified by host_id.
   """
   # Start the Process Manager
   def interested?(%HostRegistered{host_id: host_id}), do: {:start, host_id}
   def interested?(%HostRolledUp{host_id: host_id}), do: {:start, host_id}
   def interested?(%HostAddedToCluster{host_id: host_id}), do: {:start, host_id}
   def interested?(%ClusterRolledUp{snapshot: %{hosts: hosts}}), do: {:start, hosts}
+  def interested?(%DatabaseInstanceRegistered{host_id: host_id}), do: {:start, host_id}
+  def interested?(%ApplicationInstanceRegistered{host_id: host_id}), do: {:start, host_id}
+
+  def interested?(%SapSystemRolledUp{
+        snapshot: %SapSystem{
+          database: %SapSystem.Database{instances: db_instances},
+          application: %SapSystem.Application{instances: app_instances}
+        }
+      }),
+      do:
+        {:start,
+         (db_instances ++ app_instances)
+         |> Enum.map(fn %SapSystem.Instance{host_id: host_id} -> host_id end)
+         |> Enum.uniq()}
+
   # Continue the Process Manager
   def interested?(%HostDeregistrationRequested{host_id: host_id}), do: {:continue, host_id}
   def interested?(%HostRemovedFromCluster{host_id: host_id}), do: {:continue, host_id}
+  def interested?(%DatabaseInstanceDeregistered{host_id: host_id}), do: {:continue, host_id}
+  def interested?(%ApplicationInstanceDeregistered{host_id: host_id}), do: {:continue, host_id}
   # Stop the Process Manager
   def interested?(%HostDeregistered{host_id: host_id}), do: {:stop, host_id}
 
   def interested?(_event), do: false
 
-  # Deregister host that doesn't belong to any cluster
-  def handle(%DeregistrationProcessManager{cluster_id: nil}, %HostDeregistrationRequested{
-        host_id: host_id,
-        requested_at: requested_at
-      }) do
+  def handle(
+        %DeregistrationProcessManager{
+          cluster_id: nil,
+          application_instances: [],
+          database_instances: []
+        },
+        %HostDeregistrationRequested{
+          host_id: host_id,
+          requested_at: requested_at
+        }
+      ) do
     %DeregisterHost{host_id: host_id, deregistered_at: requested_at}
   end
 
-  # First step in host deregistration when host belongs to a cluster
-  def handle(%DeregistrationProcessManager{cluster_id: cluster_id}, %HostDeregistrationRequested{
-        host_id: host_id,
-        requested_at: requested_at
-      }) do
-    [
+  def handle(
+        %DeregistrationProcessManager{
+          cluster_id: nil,
```

Comment:
So now we have 40 lines function duplicated to avoid that `if` XD
I even preferred the `if` option honestly

Reply:
Or maybe some auxiliar function to flatten the app/db instances so it can be reused in the current code. At the end, that part of the code is duplicated